tasks/coverage/src/sourcemap.rs

```diff
@@ -0,0 +1,222 @@
+
+pub struct SourcemapSuite<T: Case> {
+    test_root: PathBuf,
+    test_cases: Vec<T>,
+}
+
+impl<T: Case> SourcemapSuite<T> {
+    pub fn new() -> Self {
+        Self {
+            test_root: project_root().join(FIXTURES_PATH),
+            test_cases: TestFiles::new()
+                .files()
+                .iter()
+                .filter(|file| file.file_name.contains("react"))
+                .map(|file| T::new(file.file_name.clone().into(), file.source_text.clone()))
+                .collect::<Vec<_>>(),
+        }
+    }
+}
+
+impl<T: Case> Suite<T> for SourcemapSuite<T> {
+    fn get_test_root(&self) -> &Path {
+        &self.test_root
+    }
+
+    fn save_test_cases(&mut self, tests: Vec<T>) {
+        self.test_cases.extend(tests);
+    }
+
+    fn get_test_cases(&self) -> &Vec<T> {
+        &self.test_cases
+    }
+
+    fn get_test_cases_mut(&mut self) -> &mut Vec<T> {
+        &mut self.test_cases
+    }
+
+    fn skip_test_path(&self, path: &Path) -> bool {
+        let path = path.to_string_lossy();
+        !path.contains("input.js")
+    }
+
+    fn run_coverage(&self, name: &str, _args: &crate::AppArgs) {
+        let path = project_root().join(format!("tasks/coverage/{name}.snap"));
+        let mut file = File::create(path).unwrap();
+
+        let mut tests = self.get_test_cases().iter().collect::<Vec<_>>();
+        tests.sort_by_key(|case| case.path());
+
+        for case in tests {
+            let result = case.test_result();
+            let path = case.path().to_string_lossy();
+            let result = match result {
+                TestResult::Snapshot(snapshot) => snapshot,
+                TestResult::ParseError(error, _) => error,
+                _ => {
+                    unreachable!()
+                }
+            };
+            writeln!(file, "- {path}").unwrap();
+            writeln!(file, "{result}\n\n").unwrap();
+        }
+    }
+}
+
+pub struct SourcemapCase {
+    path: PathBuf,
+    code: String,
+    source_type: SourceType,
+    result: TestResult,
+}
+
+impl SourcemapCase {
+    pub fn source_type(&self) -> SourceType {
+        self.source_type
+    }
+
+    fn generate_line_utf16_tables(content: &str) -> Vec<Vec<u16>> {
+        let mut tables = vec![];
+        let mut line_byte_offset = 0;
+        for (i, ch) in content.char_indices() {
+            match ch {
+                '\r' | '\n' | '\u{2028}' | '\u{2029}' => {
+                    // Handle Windows-specific "\r\n" newlines
+                    if ch == '\r' && content.chars().nth(i + 1) == Some('\n') {
+                        continue;
+                    }
+                    tables.push(content[line_byte_offset..i].encode_utf16().collect::<Vec<_>>());
+                    line_byte_offset = i;
+                }
+                _ => {}
+            }
+        }
+        tables.push(content[line_byte_offset..].encode_utf16().collect::<Vec<_>>());
+        tables
+    }
+
+    fn create_visualizer_text(
```

Comment:
Nice, it seems the only thing left is to improve the snapshot format.

Maybe we can move this to the codegen crate so we can re-use this in the example and playground as well?

Reply:
Not sure for this, it is a debugger format, the `https://evanw.github.io/source-map-visualization` is better to review, maybe the playground can link it if enable sourcemap.

Reply:
Oh ok, I thought we're going to copy the babel visualizer 😅